pkg/controller/user-cluster-controller-manager/constraint-syncer/wrappers_ce.go

```diff
@@ -0,0 +1,68 @@
+// +build !ee
+
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package constraintsyncer
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *reconciler) reconcile(ctx context.Context, constraint *kubermaticv1.Constraint, log *zap.SugaredLogger) error {
+	// constraint deletion
+	if constraint.DeletionTimestamp != nil {
+		if !kuberneteshelper.HasFinalizer(constraint, kubermaticapiv1.GatekeeperConstraintCleanupFinalizer) {
+			return nil
+		}
+
+		if err := r.cleanupConstraint(ctx, constraint, log); err != nil {
+			return err
+		}
+
+		oldConstraint := constraint.DeepCopy()
+		kuberneteshelper.RemoveFinalizer(constraint, kubermaticapiv1.GatekeeperConstraintCleanupFinalizer)
+		if err := r.seedClient.Patch(ctx, constraint, ctrlruntimeclient.MergeFrom(oldConstraint)); err != nil {
```

Comment:
If we are adding finalizers to the constraint from both user-cluster-controller and seed-cluster-controller we risk to end-up in a race condition like here: #6759
Maybe there is no consequence in this case, but to be safe it would be better to use optimistic locking.

Reply:
we are only adding finalizers to constraints in clsuter-ns in  user-cluster-controller

Reply:
ok, in such a case we are safe.

Reply:
yeah seed-controller adds finalizers to constraints in kubermatic ns and removes while cluster-deletion

Reply:
Well, to be more precise we are safe if one single controller is adding removing finalizers, but it should be the case, right?

Reply:
right

Reply:
although in case of cluster-deletion we set the deletion-timestamp(to avoid finalizer getting added again by user-cluster-controller) and remove the finalizer 
user-cluster-controller also has code of removing the finalizer when deletion timestamp is set, do you think it can cause problem?